cmd/kubeadm/app/util/runtime.go

```diff
@@ -0,0 +1,159 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"fmt"
+	"strings"
+
+	kubeadmapiv1alpha2 "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1alpha2"
+	utilsexec "k8s.io/utils/exec"
+)
+
+// ContainerRuntime interface
+type ContainerRuntime interface {
+	IsDocker() bool
+	IsRunning() error
+	ListContainers() ([]string, error)
+	RemoveContainers(containers []string) error
+	PullImage(image string) error
+	ImageExists(image string) (bool, error)
+}
+
+// CRIRuntime is a struct that interfaces with the CRI
+type CRIRuntime struct {
+	exec      utilsexec.Interface
+	criSocket string
+}
+
+// DockerRuntime is a struct that interfaces with the Docker
+type DockerRuntime struct {
+	exec utilsexec.Interface
+}
+
+// NewContainerRuntime sets up and returns a ContainerRuntime struct
+func NewContainerRuntime(execer utilsexec.Interface, criSocket string) (ContainerRuntime, error) {
+	var toolName string
+	var runtime ContainerRuntime
+
+	if criSocket != kubeadmapiv1alpha2.DefaultCRISocket {
+		toolName = "crictl"
+		runtime = &CRIRuntime{execer, criSocket}
+	} else {
+		toolName = "docker"
+		runtime = &DockerRuntime{execer}
+	}
+
+	if _, err := execer.LookPath(toolName); err != nil {
+		return nil, fmt.Errorf("%s is required for container runtime: %v", toolName, err)
+	}
+
+	return runtime, nil
+}
+
+// IsDocker returns true if the runtime is docker
+func (runtime *CRIRuntime) IsDocker() bool {
+	return false
+}
+
+// IsDocker returns true if the runtime is docker
+func (runtime *DockerRuntime) IsDocker() bool {
+	return true
+}
+
+// IsRunning checks if runtime is running
+func (runtime *CRIRuntime) IsRunning() error {
+	if err := runtime.exec.Command("crictl", "-r", runtime.criSocket, "info").Run(); err != nil {
+		return fmt.Errorf("container runtime is not running: %v", err)
+	}
+	return nil
+}
+
+// IsRunning checks if runtime is running
+func (runtime *DockerRuntime) IsRunning() error {
+	if err := runtime.exec.Command("docker", "info").Run(); err != nil {
+		return fmt.Errorf("container runtime is not running: %v", err)
+	}
+	return nil
+}
+
+// ListContainers lists running containers
+func (runtime *CRIRuntime) ListContainers() ([]string, error) {
+	output, err := runtime.exec.Command("crictl", "-r", runtime.criSocket, "pods", "-q").CombinedOutput()
```

Comment:
This automatically is restricted to the pods run by kubernetes? Just so we don't list anything we don't own here, right?